### PR TITLE
keras: add ability to pass live into callback

### PR DESCRIPTION
**Pull request by shcheklein**
Suggestion. If we decide it's good I can try to add it to all other integrations.

It makes this workflow (non obvious, not natural, less scalable, and not documented?): 

```python
logger = DvcLiveCallback(path="evaluation", report=None)
live = logger.dvclive
```

into:

```python
live = Live("evaluation", report=None)

# ... somewhere down the road:

.... DvcLiveCallback(dvclive = live)
```

- We need an ability to pass live logger instance or params, otherwise callback would initialize its own instance and will be producing reports for example, would ignore path, etc, etc
- It's better to pass live to callbacks (vs asking callback to give us logger): proper order of control (code looks natural, easier to test), one `live` instance could be shared across multiple callbacks, etc

### CODE REVIEW

File: src/dvclive/keras.py [hunk 1]
```diff
@@ -13,12 +14,16 @@
 
 class DvcLiveCallback(Callback):
     def __init__(
-        self, model_file=None, save_weights_only: bool = False, **kwargs
+        self,
+        model_file=None,
+        save_weights_only: bool = False,
+        dvclive: Optional[Live] = None,
```
Comment (pared) [inline src/dvclive/keras.py:20]: Should it be `dvclive` or `live`?

**Review by pared — COMMENTED** [review]
Added test, one question.

File: src/dvclive/keras.py (hunk 1 above)
Comment (shcheklein) [reply src/dvclive/keras.py:20]: I was using this name since in the __init__ itself it's using `dvclive`

Comment (shcheklein) [reply src/dvclive/keras.py:20]: I don't have any strong opinion on this tbh

**Comment by daavoo** [comment]
@shcheklein what's the use case here? 

I remember discussing this a long time ago and I think the conclusion was that it makes more sense to have a single instance (and output path) associated with the train loop. 

The primary motivation was organization-wise (separate inside/outside train loop in folders) and just overall efficiency (having a out of train loop evaluation in a separate stage avoids to recompute training when changing evaluation stuff)

**Comment by dberenbaum** [comment]
@daavoo Maybe you are thinking of https://github.com/iterative/dvclive/issues/207?

**Comment by shcheklein** [comment]
@daavoo use case is when you want to log certain things besides what's being tracked inside one of the callbacks, so you need to have access to the `live` instance.

This is the way it can be solved:

https://github.com/shcheklein/hackathon/blob/universe/train.py#L55-L56

But I think this is wrong to do it that way (wrong control order). I've outlined a few reasons for that above.

**Comment by shcheklein** [comment]
Alternatively maybe `live` instance can be global (similar to Python logger?) and callbacks can rely on that instance?

**Comment by shcheklein** [comment]
@daavoo @dberenbaum yes, #207 and #289 are the same / very related. I hit this bump independently.

**Comment by dberenbaum** [comment]
> Alternatively maybe `live` instance can be global (similar to Python logger?) and callbacks can rely on that instance?

I think it used to be global but we moved away from it in case users want multiple live instances (for example, they train multiple models). IMO optionally passing live into the callback (as this PR does) is the most flexible since you can choose whether you want to keep everything in one instance or not.

**Review by daavoo — APPROVED** [review]

**Comment by daavoo** [comment]
> @daavoo use case is when you want to log certain things besides what's being tracked inside one of the callbacks, so you need to have access to the live instance.
> 
> This is the way it can be solved:
> 
> https://github.com/shcheklein/hackathon/blob/universe/train.py#L55-L56
> 
> But I think this is wrong to do it that way (wrong control order). I've outlined a few reasons for that above.

I understand the low-level motivation, my question was about the high-level use case, which in the linked example appears to be something like: "log the best value of each metric so that those are the values shown in the table".

I am ok with the P.R. but my feeling still is that the use cases for this are either workarounds of existing dvclive limitations or cases where would be a better practice to have a separate stage and/or output folder.

**Comment by shcheklein** [comment]
> but my feeling still is that the use cases for this are either workarounds of existing dvclive limitations or cases where would be a better practice to have a separate stage and/or output folder.

Yes, I think I understand your feelings. My intuition though that this kind of extensibility / flexibility will be still needed (it will be very hard to fit / cover everything with callbacks / higher level APIs). Also, tbh, I don't see that much harm exposing this. We can keep learning from the way people using it and migrate scenarios inside.

I'll try to expand to all other frameworks / add docs if needed.

**Comment by pared** [comment]
I think that this approach might be useful if we don't want to log everything on each step:
let's say that we want a callback to log training metrics, but only want to dump something "big" (confusion matrix for example) at the end of the training.
